server/src/main/java/org/apache/druid/metadata/SqlSegmentsMetadataQuery.java

```diff
@@ -233,12 +233,24 @@ private CloseableIterator<DataSegment> retrieveSegments(
             )
         );
 
-        if (i == intervals.size() - 1) {
-          sb.append(")");
-        } else {
+        // Add a special check for a segment which have one end at eternity and the other at some finite value. Since
+        // we are using string comparison, a segment with this start or end will not be returned otherwise.
+        if (matchMode.equals(IntervalMode.OVERLAPS)) {
+          sb.append(StringUtils.format(" OR (start = '%s' AND \"end\" != '%s' AND \"end\" > :start%d)", Intervals.ETERNITY.getStart(), Intervals.ETERNITY.getEnd(), i));
```

Comment:
It would be cool if we add the the before and after sample sql statements in the PR desc. Would be easy to review the change that way.

Reply:
Good idea! Done